### PR TITLE
Add protoc's search path to the plugin API.

### PR DESCRIPTION
**Pull request by PetteriAimonen**
ProtoC automatically searches for include files in the path
given by -I command line option. Only the file name is given
to the code generator plugins, not the full path.

This is fine in itself, but in order to find the actual file
for further analysis (e.g. custom formatting, comment or options processing)
in a generator plugin, it needs to know the search path used
by protoc. This commit adds that informaion to CodeGeneratorRequest.

### CODE REVIEW

**Comment by PetteriAimonen** [comment]
I signed the CLA.

**Comment by PetteriAimonen** [comment]
Hmm, tried to add email to CLA, correct now?

**Comment by xfxyjwf** [comment]
The import path is only used by protoc when searching for imports. When the descriptor information is presented to plugins, the info for all imported files is already there (in FileDescriptorSet). Could you elaborate what else you want from the import paths?

**Comment by PetteriAimonen** [comment]
@xfxyjwf In nanopb's case, we have a few extra settings that allow more embedded-friendly behaviour, such as static allocations by setting the maximum size for arrays. These can be set using options inside the .proto file, but several users do not want to modify their .proto files for the purposes of just one library. Thus nanopb also supports separate .options files, which it tries to look up in the same place as the .proto file. But when run through protoc, it is not possible to accurately know the path of the .proto file.

It seems to me that the same problem exists with the SourceCodeInfo already provided by protoc: in order to take full advantage of it, the original .proto file must be available to the plugin.

I guess same effect could be achieved by adding an absolute file path to each FileDescriptorProto, but that is a more invasive change than the one proposed.

**Comment by PetteriAimonen** [comment]
Any updates on this? This seems like a low-impact change on protobuf end, and I kind-of need it on the nanopb end. I really wouldn't want to have to maintain a custom fork of protoc.

File: src/google/protobuf/compiler/plugin.proto [hunk 1]
```diff
@@ -63,6 +63,9 @@ message CodeGeneratorRequest {
   // The generator parameter passed on the command-line.
   optional string parameter = 2;
 
+  // Search path used by protoc (-I command line option).
+  repeated string search_path = 3;
```
Comment (xfxyjwf) [inline src/google/protobuf/compiler/plugin.proto]: Use "proto_path" to be consistent with the command line flag name?

**Comment by xfxyjwf** [comment]
This change looks good to me. @pherl , could you take a look at this plugin interface change as well?

File: src/google/protobuf/compiler/plugin.proto (hunk 1 above)
Comment (PetteriAimonen) [inline src/google/protobuf/compiler/plugin.proto]: Renamed to "proto_path" per suggestion.

**Comment by PetteriAimonen** [comment]
Any status updates on this?

**Comment by pherl** [comment]
Ideally would you also need the PWD in which protoc is invoked?

**Comment by PetteriAimonen** [comment]
As far as I know, the current working directory for a plugin is the same as the current working directory for protoc itself? So it does not have to be separately passed.

**Comment by pherl** [comment]
-I can also be used in a fully qualified mapping way: -I{virtual_path}=real_path, e.g. -Ifoo.proto=xxx/xxx/foo.proto
Those should be covered, and we need some tests for it.

**Comment by PetteriAimonen** [comment]
Looks like I missed the full mapping part of the -I option. I'll have to remake the patch when I have time.

**Comment by haberman** [comment]
Feel free to reopen if you want to rework the patch later. We haven't had any other requests for this that I know of.

**Pull request closed without merging**